rfd/00xx-kubernetes-access-cluster-discovery.md

```diff
@@ -0,0 +1,428 @@
+- cluster:
+    certificate-authority: ca.crt
+    server: https://127.0.0.1:52181
+  name: minikube
+contexts:
+- context:
+    cluster: minikube
+    namespace: default
+    user: minikube
+  name: minikube
+kind: Config
+users:
+- name: minikube
+  user:
+    client-certificate: client.crt
+    client-key: client.key
+```
+
+- `clusters`: defines the `cluster` API and certificate authority.
+- `users`: describes `user` credentials.
+- `contexts`: links a `cluster` and `user` definition.
+
+In Teleport, the `cluster` details are used to create the cluster config and the `users` credentials are stored in credentials map. This map links the cluster name into the credentials. The discovery mechanism will leverage this behavior and dynamically enroll clusters using the same philosophy but without the necessity of writing or editing the kubeconfig file.
+
+For each cluster to enroll, due to limitations described in the [EKS Discovery and IAM](#aws-eks-discovery-and-iam) section, Teleport cannot create access to itself, so the operator has to create the following resources for each cluster he pretends that Teleport enrolls.
+
+The first resource to be created is the `ClusterRole` that grants minimum RBAC permissions to get pods and impersonate groups, users and service accounts. It is possible to run map Teleport role into `system:masters` but for security reasons that is discouraged.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: teleport-role
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - users
+  - groups
+  - serviceaccounts
+  verbs:
+  - impersonate
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  verbs:
+  - get
+- apiGroups:
+  - "authorization.k8s.io"
+  resources:
+  - selfsubjectaccessreviews
+  - selfsubjectrulesreviews
+  verbs:
+  - create
+```
+
+The second step is to create the cluster role binding to map the cluster role into the Kubernetes Teleport group.
+
+```yaml
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: teleport-role-binding
+subjects:
+- kind: Group
+  name: system:teleport
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Finally, the operator has to map the Teleport IAM role into the Kubernetes group in the `configmap/aws-auth`:
+
+```yaml
+apiVersion: v1
+data:
+  mapRoles: |
+    - groups:
+      - system:teleport
+      rolearn: arn:aws:iam::222222222222:role/teleport-role
+      username: system:teleport
+...
+```
+
+At this point, Teleport discovery can run the cluster discovery mechanism every minute and monitor if new clusters appeared or if any was deleted. When a new cluster is found, Teleport checks if it has access to it by generating short-lived credentials and doing requests to the API. If the request is successful, it registers the cluster by creating the cluster config and stores the short-lived credentials in the credentials map. Teleport will also import, for each discovered cluster, the EKS cluster tags as labels.
```

Comment:
> stores the short-lived credentials in the credentials map

The credential mentioned here is the IAM auth token, correct? How is this map updated when the token expires?

Reply:
Yes, it is a special IAM auth token that grants access to a specific EKS cluster (The STS request is special and is handed by [aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator).
The credentials are refreshed independently if they are near expiration. The expiration date is returned from AWS API and I refresh them once they pass the 1/2*TTL and TTL is defined at IAM role level so the user can control it. 

In my implementation, I changed the `credentials` map to receive an interface and I created two structures that satisfy the interface.

- `staticCredentials`: credentials that never expire and are imported either from KubeConfig or in cluster SA
- `dynamicCredentials`: credentials that have TTL and must be renewed for each Cluster. Once the credentials are renewed the `dynamicCredentials` are updated with the new TOKEN (AWS EKS)